src/commands/validate.js

```diff
@@ -40,64 +43,52 @@ function getOverallErrors(variations = [], components = [], log, warn, error) {
-    yargs.option('variations', {
-      demandOption: true,
-      requiresArg: true,
-    });
-    yargs.option('components', {
-      demandOption: true,
-      requiresArg: true,
-    });
-  }
-  yargs.positional('variations', {});
-
-  if (all && projectCount <= 0) {
-    throw chalk.red(`\`--all\` requires a non-empty “projects” config`);
+  if (all && allProjectNames.length <= 0) {
+    throw chalk.red('`--all` requires a non-empty “projects” config');
   }
   if (all && project) {
-    throw chalk.red(`\`--all\` and \`--project\` are mutually exclusive`);
+    throw chalk.red('`--all` and `--project` are mutually exclusive');
   }
   if (project && !has(projects, project)) {
     throw chalk.red(`Project "${project}" missing from “projects” config`);
   }
-  if (project) {
-    yargs.config(projects[project]);
+
+  if (projects) {
+    validateProjects(projects, allProjectNames, 'in the “projects” config');
+  } else {
+    validateProject(config);
   }
 };
-exports.handler = ({ variations, components, all, projects }) => {
-  if (all) {
-    const exitCodes = Object.entries(projects).map(([
-      project,
-      { variations, components, require: requires },
-    ]) => {
-      if (requires) { requireFiles(requires); }
-      try {
-        return getOverallErrors(
-          globToFiles(variations),
-          globToFiles(components),
-          x => console.log(`${chalk.inverse(chalk.blue(`Project “${project}”`))}: ${x}`),
-          x => console.warn(`${chalk.inverse(chalk.yellow(`Project “${project}”`))}: ${x}`),
-          x => console.error(`${chalk.inverse(chalk.red(`Project “${project}”`))}: ${x}`),
-        );
-      } catch (e) {
-        console.error(`${chalk.inverse(chalk.red(`Project “${project}”`))}: ${e.message}`);
-        return 1;
-      }
-    });
-    process.exit(Math.max(...exitCodes));
-  }
 
-  const exitCode = getOverallErrors(
+exports.handler = (config) => {
+  const { projects, projectNames } = normalizeConfig(config);
+
+  const exitCodes = [];
+
+  forEachProject(projects, projectNames, (project, {
     variations,
     components,
-    x => console.log(x),
-    x => console.warn(x),
-    x => console.error(x),
-  );
-  process.exit(exitCode);
+    require: requires,
+  }) => {
+    if (requires) { requireFiles(requires); }
+    try {
+      const exitCode = getOverallErrors(
+        globToFiles(variations),
+        globToFiles(components),
+        x => console.log(`${chalk.inverse(chalk.blue(`Project “${project}”`))}: ${x}`),
+        x => console.warn(`${chalk.inverse(chalk.yellow(`Project “${project}”`))}: ${x}`),
+        x => console.error(`${chalk.inverse(chalk.red(`Project “${project}”`))}: ${x}`),
+      );
+      exitCodes.push(exitCode);
+    } catch (e) {
+      console.error(`${chalk.inverse(chalk.red(`Project “${project}”`))}: ${e.message}`);
+      exitCodes.push(1);
+    }
+  });
+
+  process.exit(Math.max(...exitCodes));
```

Comment:
If we get any non-zero and non-one exit codes, it might be worth dropping a log that lists all of the exit codes and project they are associated with. Or maybe that's what the `x` callbacks up there do?

Regardless, I wonder if this should be changed to something that exits with a 1 if any are non-zero? Or maybe special logic to pass through exit codes with special meaning? http://tldp.org/LDP/abs/html/exitcodes.html Seems weird that it would just pick the highest one. I'm probably overthinking this.

Reply:
Since I’m determining all the exit codes, i have a loose convention of “the highest one is the worst” - i agree that the specific code isn’t providing much meaning, but I’m hoping it helps me debug.